pkg/cloud/services/networking/network.go

```diff
@@ -369,3 +370,28 @@ func GetSubnetsByFilter(networkClient *gophercloud.ServiceClient, opts subnets.L
 	}
 	return snets, nil
 }
+
+func (s *Service) DeleteOrphanedPorts() error {
+	portList, err := ports.List(s.client, ports.ListOpts{ProjectID: s.projectID}).AllPages()
+	if err != nil {
+		return fmt.Errorf("error listing ports: %v", err)
+	}
+
+	orphanedPorts, err := ports.ExtractPorts(portList)
+	if err != nil {
+		return fmt.Errorf("error extracting ports: %v", err)
+	}
+
+	for _, orphanedPort := range orphanedPorts {
+		if orphanedPort.DeviceOwner != "" {
```

Comment:
We shouldn't delete anything which is not created by ClusterAPI. This statement would lead to deletion of any ports which were not created by ClusterAPI but were created for other purposes (either manually or via some other automation).

One common use case for this: In cases where a loadbalancer is not available (or the user does not use if for whatever reasons), and instead a VIP (Virtual IP) along with `keepalived` is used for load balancing, then an openstack port is created (to reserve the IP and to create proper routes), but is not attached to any VM. Since this openstack port is not attached to any VM, it has `device_owner == ""`.

The `DeleteOrphanedPorts` function would delete this port unnecessarily and would cause service disruption elsewhere.

Reply:
@rustycl0ck Understood. So I assume there is no real way to cleanup those ports safely. I'll drop this for now from the e2e tests. Not sure how we can get rid of the problems during cluster deletion though. I'll try if I can get the e2e tests clean without this change but I'm not sure if it's possible

Reply:
I'm not entirely sure about the general statement "we shouldn't delete anything which is not created by ClusterAPI". In my opinion there are cases where it's totally valid. One case is that the cluster as a whole should be deleted and somebody has the clean up the loadbalancers which has been created by the cloud provider openstack. This is also done in CAPA and the ccm itself is unable to delete them as it doesn't exist anymore.

Are those service interruptions really an issue when the whole cluster is deleted?

Reply:
My use case was referring to a setup that I worked on, where a single openstack network (IPv6) was being used by multiple teams/products. While I was trying to use CAPI to setup a k8s cluster there, a few other teams had used the VIP approach for making their applications HA (since a loadbalancer wasn't available). Just FYI, the other products weren't k8s based, but legacy VM based systems/appliccations.

#723 was caused at the time of cluster creation where the port is created first, but if the image (or the ssh key or the flavor) does not exist, server creation would error out and the port would be left behind without any cleanup on server creation failure. This issue seems to have been addressed in #705 correctly (where a rollback is performed if the server creation failed). Similarly, we should try to fix the root cause instead of performing a general cleanup.

I think the correct way to fix this would be to figure out why this error occurred in the first place.
From the following code block: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/pkg/cloud/services/compute/instance.go#L631-L648, I don't see how/why the port would have been left behind while the server got deleted successfully. If we can find the root cause, we can fix the problem in a cleaner way.

Reply:
@rustycl0ck Thx providing some more details :). 

I think overall you're right. The better approach would be to cleanup the ports wherever they orphaned. I'm currently testing if this PR also works without the `DeleteOrphanedPorts` func: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/816
I assume this will be successful and then I'll drop the func on the current PR.

Out of curiosity (I'm not sure what is possible with OpenStack). Did you have:
* a separate project and where using the same project and network as the other teams or
* have you been using the same network but separate projects?

Reply:
It was a single project and same network, because all the applications (k8s hosted as well as legacy VM based) had to interact with each other (on a private network internally).

Reply:
As I feared the tests are now failing because they are timing out because the cluster deletion does not work: https://prow.k8s.io/log?container=test&id=1376784008490258432&job=pull-cluster-api-provider-openstack-e2e-test

I would suggest merging the PR anyway without the DeleteOrphanedPort func. I would rather have those tests with a known issue on master and unblock a lot of other PRs instead of having to rely only on the conformance tests.Those tests are not mandatory yet and I try to find the bug in a follow-up PR.

What do you think?
(@jichenjc )

Reply:
Got the exact results. Looks like the error only happens when server are not successfully created right now. I'll check if I can just fix this...

Reply:
Fixed it here: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/798/commits/2e012c6e8fbbf51a844f9f514c62544c5ff998b1
Local tests were green. I'll squash and re-run the tests on this PR